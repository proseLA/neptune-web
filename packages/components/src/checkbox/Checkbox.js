import classNames from 'classnames';
import PropTypes from 'prop-types';

import CheckboxButton from '../checkboxButton';

const Checkbox = ({
  id,
  checked,
  required,
  disabled,
  readOnly,
  label,
  secondary,
  onChange,
  onFocus,
  onBlur,
}) => {
  const hasError = required && !disabled && !readOnly && !checked;

  const classList = classNames('np-checkbox', {
    checkbox: true,
    'checkbox-lg': secondary,
    'has-error': hasError,
    disabled,
  });

  return (
    <div id={id} className={classList}>
      {/* eslint-disable jsx-a11y/label-has-for */}
      <label>
        <CheckboxButton
          className={classNames('p-r-2', { 'has-error': hasError })}
          checked={checked}
          disabled={disabled || readOnly}
          onFocus={onFocus}
          onChange={() => onChange(!checked)}
          onBlur={onBlur}
        />
        <span className="np-checkbox__text">
          {label}
          {required && '*'}
          {secondary && <small>{secondary}</small>}
        </span>
      </label>
    </div>
  );
};

Checkbox.propTypes = {
  id: PropTypes.string,
  checked: PropTypes.bool,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  label: PropTypes.node.isRequired,
  secondary: PropTypes.string,
  onFocus: PropTypes.func,
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func,
};

Checkbox.defaultProps = {
  id: null,
  checked: false,
  required: false,
  disabled: false,
  readOnly: false,
  secondary: null,
  onFocus: null,
  onBlur: null,
};

export default Checkbox;
