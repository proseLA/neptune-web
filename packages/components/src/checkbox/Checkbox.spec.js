import '@testing-library/jest-dom';

import { render, fireEvent, screen } from '@testing-library/react';

import Checkbox from '.';

describe('Checkbox', () => {
  let props;

  beforeEach(() => {
    props = {
      label: <b>hello</b>,
      onChange: jest.fn(),
    };
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders the given label', () => {
    const { container } = render(<Checkbox {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('is enabled by default', () => {
    const { container } = render(<Checkbox {...props} />);
    expect(getCheckboxContainer(container)).not.toHaveClass('disabled');
  });

  it('does not render secondary text', () => {
    const { container } = render(<Checkbox {...props} />);
    expect(container.querySelector('small')).not.toBeInTheDocument();
  });

  it('calls change handler with new checked value when checkbox button is clicked', () => {
    const onChange = jest.fn();

    const { rerender } = render(<Checkbox {...props} onChange={onChange} />);

    expect(onChange).not.toHaveBeenCalled();
    fireEvent.click(getLabel());
    expect(onChange).toHaveBeenCalledWith(true);

    rerender(<Checkbox {...props} checked onChange={onChange} />);

    fireEvent.click(getLabel());
    expect(onChange).toHaveBeenCalledWith(false);
  });

  it('does not call change handler on checkbox button click when disabled', () => {
    const onChange = jest.fn();

    render(<Checkbox {...props} disabled onChange={onChange} />);

    fireEvent.click(getLabel());
    expect(onChange).not.toHaveBeenCalled();
  });

  it('does not call change handler on checkbox button click when readOnly', () => {
    const onChange = jest.fn();

    render(<Checkbox {...props} readOnly onChange={onChange} />);

    fireEvent.click(getLabel());
    expect(onChange).not.toHaveBeenCalled();
  });

  it('has disabled class when the disabled prop is true', () => {
    const { container } = render(<Checkbox {...props} disabled />);
    expect(getCheckboxContainer(container).firstChild).toHaveClass('disabled');
  });

  it('has error class and passes it to checkbox button when required and not checked', () => {
    const { container, rerender } = render(<Checkbox {...props} required checked />);
    expect(getCheckboxContainer(container)).not.toHaveClass('has-error');

    rerender(<Checkbox {...props} required />);

    expect(getCheckboxContainer(container)).toHaveClass('has-error');
  });

  it('ignores required if disabled or readOnly', () => {
    const { container, rerender } = render(<Checkbox {...props} required disabled />);
    expect(getCheckboxContainer(container)).not.toHaveClass('has-error');

    rerender(<Checkbox {...props} required readOnly />);
    expect(getCheckboxContainer(container)).not.toHaveClass('has-error');
  });

  it('has an asterisk after the label when required', () => {
    render(<Checkbox {...props} required />);

    expect(screen.getByText('*')).toBeInTheDocument();
  });

  it(`doesn't passes checked to checkbox button`, () => {
    render(<Checkbox {...props} />);
    expect(getCheckbox()).not.toBeChecked();
  });

  it('passes checked to checkbox button', () => {
    render(<Checkbox {...props} checked />);
    expect(getCheckbox()).toHaveAttribute('checked');
  });

  it('passes disabled to checkbox button', () => {
    const { rerender } = render(<Checkbox {...props} />);
    expect(getCheckbox()).toBeEnabled();

    rerender(<Checkbox {...props} disabled />);
    expect(getCheckbox()).toHaveAttribute('disabled');
  });

  it('disables checkbox button if readOnly', () => {
    const { rerender } = render(<Checkbox {...props} />);

    expect(getCheckbox()).toBeEnabled();

    rerender(<Checkbox {...props} readOnly />);
    expect(getCheckbox()).toHaveAttribute('disabled');
  });

  it('displays secondary text when supplied', () => {
    const { container } = render(<Checkbox {...props} secondary="secondary text" />);
    expect(screen.getByText('secondary text')).toBeInTheDocument();
    expect(getCheckboxContainer(container)).toHaveClass('checkbox-lg');
  });

  const getCheckboxContainer = (container) => container.querySelector('.np-checkbox');
  const getLabel = () => screen.getByText('hello').parentElement;
  const getCheckbox = () => screen.getByRole('checkbox');
});
