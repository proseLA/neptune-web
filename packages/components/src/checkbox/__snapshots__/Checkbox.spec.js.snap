// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders the given label 1`] = `
<div>
  <div
    class="np-checkbox checkbox"
  >
    <label
      class=""
    >
      <span
        class="np-checkbox-button p-r-2"
      >
        <input
          type="checkbox"
        />
        <span
          class="tw-checkbox-button"
        >
          <span
            class="tw-checkbox-check"
          />
        </span>
      </span>
      <span
        class="np-checkbox__text"
      >
        <b>
          hello
        </b>
      </span>
    </label>
  </div>
</div>
`;
